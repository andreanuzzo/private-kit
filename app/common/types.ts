import { Posix } from '../helpers/dateTimeUtils';

export type LocalDataPoint = {
  latitude: number;
  longitude: number;
  time: Posix;
  hashes: string[];
};

export type Coordinates = {
  latitude: number;
  longitude: number;
};

export type ApiHealthcareAuthority = {
  name: string;
  bounds: {
    ne: Coordinates;
    sw: Coordinates;
  };
  org_id: string;
  cursor_url: string;
  public_api: string;
};

export type HealthcareAuthority = ApiHealthcareAuthority & {
  internal_id: string;
};
