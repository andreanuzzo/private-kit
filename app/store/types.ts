import { store } from './index';
import rootReducer from './reducers/rootReducer';
import { StateType } from 'typesafe-actions';
import { ThunkAction } from 'redux-thunk';
import { Action } from 'redux';
// Our internal store representation is the same as the API representation
export type { HealthcareAuthority } from '../common/types';

export type RootState = ReturnType<typeof rootReducer>;
export type Store = StateType<typeof store>;

// These will appear in redux debugger, so string names are helpful
export enum ApiStatus {
  INITIAL = 'INITIAL',
  STARTED = 'STARTED',
  SUCCESS = 'SUCCESS',
  FAILURE = 'FAILURE',
}

export type ApiRequest = {
  status: ApiStatus; // This allows components to show states based on api status
  errorMessage: string | null; // This is only for redux debugging. Store as a string for the safety in the store.
};

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export enum FeatureFlagOption {
  IMPORT_LOCATIONS_GOOGLE = 'IMPORT_LOCATIONS_GOOGLE',
  IMPORT_LOCATIONS_JSON_URL = 'IMPORT_LOCATIONS_JSON_URL',
  CUSTOM_URL = 'CUSTOM_URL',
  DOWNLOAD_LOCALLY = 'DOWNLOAD_LOCALLY',
  DEV_LANGUAGES = 'DEV_LANGUAGES',
  BYPASS_EXPORT_API = 'BYPASS_EXPORT_API',
  MOCK_EXPOSURE = 'MOCK_EXPOSURE',
}
