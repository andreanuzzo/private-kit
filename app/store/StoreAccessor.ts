import { Store } from 'redux';

/*
 * This accessor is added for the SOLE purpose of avoiding cyclic imports from
 * the IntersectService to the Intersect helper and again to the store. The
 * accessor serves as a bridge for the consumers.
 * */
class StoreAccessor {
  store: Store | null = null;

  setStore = (store: Store | null) => {
    this.store = store;
  };

  getStore = () => {
    return this.store;
  };
}

const singleton = new StoreAccessor();

export default singleton;
