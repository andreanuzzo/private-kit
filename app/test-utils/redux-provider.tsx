// test-utils.js
import React, { ReactElement, ReactNode } from 'react';
import {
  render as rtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react-native';
import { Store, createStore } from 'redux';
import { Provider, DefaultRootState } from 'react-redux';
import rootReducer from '../store/reducers/rootReducer';

type ReduxRenderOptions = RenderOptions & {
  initialState: DefaultRootState;
  store?: Store;
};

type ReduxProviderProps = {
  children?: ReactNode;
};

function render(
  ui: ReactElement<unknown>,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...rest
  }: ReduxRenderOptions,
): RenderResult {
  function ReduxProvider({ children }: ReduxProviderProps) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: ReduxProvider, ...rest });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
