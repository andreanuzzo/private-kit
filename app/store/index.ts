import AsyncStorage from '@react-native-community/async-storage';
import { createImmutableStateInvariantMiddleware } from '@reduxjs/toolkit';
import { applyMiddleware, createStore } from 'redux';
import { createMigrate, persistReducer, persistStore } from 'redux-persist';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';

import StoreAccessor from './StoreAccessor';
import onChangedSelectedHealthAuthorities from './middleware/onChangedSelectedHealthAuthorities';

import migrations from './migrations';
import rootReducer from './reducers/rootReducer';

const enhancers = composeWithDevTools(
  applyMiddleware(
    thunk,
    createImmutableStateInvariantMiddleware(),
    onChangedSelectedHealthAuthorities,
  ),
);

export const STORE_VERSION = 2;
const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  version: STORE_VERSION,
  // add keys of reducers to ignore:
  blacklist: [],
  migrate: createMigrate(migrations),
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(persistedReducer, {}, enhancers);
StoreAccessor.setStore(store);
export const persistor = persistStore(store);
