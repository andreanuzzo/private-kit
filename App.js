import React, { useEffect } from 'react';
import SplashScreen from 'react-native-splash-screen';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import 'text-encoding-polyfill';
import 'array-flat-polyfill';

import { Entry } from './app/Entry';
import { TracingStrategyProvider } from './app/TracingStrategyContext';
import VersionCheckService from './app/services/VersionCheckService';
import { store, persistor } from './app/store';
import gpsStrategy from './app/gps';
import BackgroundTaskService from './app/services/BackgroundTaskService';

// For snapshot testing. In tests, we provide a mock store wrapper if needed.
export const UnconnectedApp = () => (
  <TracingStrategyProvider strategy={gpsStrategy}>
    <Entry />
  </TracingStrategyProvider>
);

const App = () => {
  useEffect(() => {
    SplashScreen.hide();
    VersionCheckService.start();
    BackgroundTaskService.start();
  }, []);

  return (
    <Provider store={store}>
      <PersistGate persistor={persistor}>
        <UnconnectedApp />
      </PersistGate>
    </Provider>
  );
};

export default App;
