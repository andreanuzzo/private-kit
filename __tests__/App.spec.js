/**
 * @format
 */

import 'react-native';
import 'isomorphic-fetch';

import { render } from '../app/test-utils/redux-provider';
import React from 'react';

import { UnconnectedApp } from '../App';
import { FeatureFlagOption } from '../app/store/types';

jest.mock('../app/Entry', () => ({ Entry: 'Entry' }));

// We're using the unconnected App, so omit creating a store.
jest.mock('../app/store', () => ({ createPersistedStore: () => ({}) }));

it('renders correctly', () => {
  const { asJSON } = render(<UnconnectedApp />, {
    initialState: {
      featureFlags: { flags: { [FeatureFlagOption.MOCK_EXPOSURE]: false } },
    },
  });

  expect(asJSON()).toMatchSnapshot();
});
