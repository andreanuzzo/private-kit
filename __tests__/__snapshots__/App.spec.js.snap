// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Entry />
</View>
`;
